### PR TITLE
Created DB migration and model for Questions

### PR DESCRIPTION
**Pull request by maybee16**
http://outsourcing.nat.tepkom.ru:20080/redmine/issues/1128

### CODE REVIEW

File: src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+          HasRealTimeResult = table.Column<bool>(nullable: false),
+          IsAnonymous = table.Column<bool>(nullable: false),
+          IsRevoteAvailable = table.Column<bool>(nullable: false),
+          IsObligatory = table.Column<bool>(nullable: false),
+          IsPrivate = table.Column<bool>(nullable: false),
+          HasMultipleChoice = table.Column<bool>(nullable: false),
+          HasCustomOptions = table.Column<bool>(nullable: false),
+          CreatedBy = table.Column<Guid>(nullable: false),
+          CreatedAtUTC = table.Column<DateTime>(nullable: false),
+          ModifiedBy = table.Column<Guid>(nullable: false),
```
Comment (AlfaRedDog) [inline src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs]: "ModifiedBy" property can be null

File: src/SurveyService.Models.Db/DbQuestion.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    public bool HasRealTimeResult { get; set; }
+    public bool IsAnonymous { get; set; }
+    public bool IsRevoteAvailable { get; set; }
+    public bool IsObligatory { get; set; }
+    public bool IsPrivate { get; set; }
+    public bool HasMultipleChoice { get; set; }
+    public bool HasCustomOptions { get; set; }
+    public Guid CreatedBy { get; set; }
+    public DateTime CreatedAtUTC { get; set; }
+    public Guid ModifiedBy { get; set; }
```
Comment (AlfaRedDog) [inline src/SurveyService.Models.Db/DbQuestion.cs]: "ModifiedBy" property can be null

File: src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs (hunk 1 above)
Comment (maybee16) [reply src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs]: Done

File: src/SurveyService.Models.Db/DbQuestion.cs (hunk 2 above)
Comment (maybee16) [reply src/SurveyService.Models.Db/DbQuestion.cs]: Done

**Review by AlfaRedDog — APPROVED** [review]

File: src/SurveyService.Models.Db/DbQuestion.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+      builder
+          .HasKey(q => q.Id);
+
+      builder
+          .Property(q => q.Content)
+          .IsRequired();
+
+      builder
+          .HasMany(o => o.Options)
+          .WithOne(o => o.Questions);
```
Comment (Levliv) [inline src/SurveyService.Models.Db/DbQuestion.cs]: .HasMany(q => q.Options)
.WithOne(o => o.Question);

File: src/SurveyService.Models.Db/DbQuestion.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+      builder
+          .HasOne(g => g.Group)
+          .WithMany(o => o.Questions);
```
Comment (Levliv) [inline src/SurveyService.Models.Db/DbQuestion.cs]: .HasOne(q => q.Group)
.WithMany(g => g.Questions);

File: src/SurveyService.Models.Db/DbQuestion.cs (hunk 3 above)
Comment (maybee16) [reply src/SurveyService.Models.Db/DbQuestion.cs]: Done

File: src/SurveyService.Models.Db/DbQuestion.cs (hunk 4 above)
Comment (maybee16) [reply src/SurveyService.Models.Db/DbQuestion.cs]: Done

File: src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
+﻿using LT.DigitalOffice.SurveyService.Models.Db;
+using Microsoft.EntityFrameworkCore.Infrastructure;
+using Microsoft.EntityFrameworkCore.Migrations;
+using System;
+
+namespace LT.DigitalOffice.SurveyService.Data.Provider.MsSql.Ef.Migrations;
+
+[DbContext(typeof(SurveyServiceDbContext))]
+[Migration("20220826180800_InitialQuestionsTable")]
```
Comment (vorobevamv) [inline src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs]: While service is not in release, better have one migration for initial tables - it is easier to fix and read; other tables can just would be added to this migration.
So, please, change name to "InitialCreate"

File: src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+class InitialQuestionsTable : Migration
```
Comment (vorobevamv) [inline src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs]: "InitialCreate"

File: src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        Id = table.Column<Guid>(nullable: false),
+        GroupId = table.Column<Guid>(nullable: true),
+        Content = table.Column<string>(nullable: false),
+        Deadline = table.Column<DateTime>(nullable: true),
+        HasRealTimeResult = table.Column<bool>(nullable: false),
+        IsAnonymous = table.Column<bool>(nullable: false),
+        IsRevoteAvailable = table.Column<bool>(nullable: false),
+        IsObligatory = table.Column<bool>(nullable: false),
+        IsPrivate = table.Column<bool>(nullable: false),
+        HasMultipleChoice = table.Column<bool>(nullable: false),
```
Comment (vorobevamv) [inline src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs]: HasMultiplyChoice

File: src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        HasCustomOptions = table.Column<bool>(nullable: false),
+        CreatedBy = table.Column<Guid>(nullable: false),
+        CreatedAtUTC = table.Column<DateTime>(nullable: false),
```
Comment (vorobevamv) [inline src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs]: CreatedAtUtc

File: src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        ModifiedBy = table.Column<Guid>(nullable: true),
+        ModifiedAtUTC = table.Column<DateTime>(nullable: true),
```
Comment (vorobevamv) [inline src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs]: ModifiedAtUtc

File: src/SurveyService.Data.Provider.MsSql.Ef/SurveyService.Data.Provider.MsSql.Ef.csproj [hunk 10]
```diff
@@ -7,8 +7,4 @@
     <ProjectReference Include="..\SurveyService.Models.Dto\SurveyService.Models.Dto.csproj" />
   </ItemGroup>
 
-  <ItemGroup>
-    <Folder Include="Migrations\" />
-  </ItemGroup>
```
Comment (vorobevamv) [inline src/SurveyService.Data.Provider.MsSql.Ef/SurveyService.Data.Provider.MsSql.Ef.csproj:12]: Have you deleted this (lines 10-12) by hands or it've happened automatically?

File: src/SurveyService.Models.Db/DbQuestion.cs [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+  public Guid Id { get; set; }
+  public Guid? GroupId { get; set; }
+  public string Content { get; set; }
+  public DateTime? Deadline { get; set; }
+  public bool HasRealTimeResult { get; set; }
+  public bool IsAnonymous { get; set; }
+  public bool IsRevoteAvailable { get; set; }
+  public bool IsObligatory { get; set; }
+  public bool IsPrivate { get; set; }
+  public bool HasMultipleChoice { get; set; }
```
Comment (vorobevamv) [inline src/SurveyService.Models.Db/DbQuestion.cs:21]: HasMultiplyChoice

File: src/SurveyService.Models.Db/DbQuestion.cs [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+  public bool HasCustomOptions { get; set; }
+  public Guid CreatedBy { get; set; }
+  public DateTime CreatedAtUTC { get; set; }
```
Comment (vorobevamv) [inline src/SurveyService.Models.Db/DbQuestion.cs]: CreatedAtUtc

File: src/SurveyService.Models.Db/DbQuestion.cs [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+  public Guid? ModifiedBy { get; set; }
+  public DateTime? ModifiedAtUTC { get; set; }
```
Comment (vorobevamv) [inline src/SurveyService.Models.Db/DbQuestion.cs]: ModifiedAtUtc

File: src/SurveyService.Models.Db/DbQuestion.cs [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+    builder
+        .HasKey(q => q.Id);
+
+    builder
+        .Property(q => q.Content)
+        .IsRequired();
+
+    builder
+        .HasMany(q => q.Options)
+        .WithOne(o => o.Questions);
```
Comment (Levliv) [inline src/SurveyService.Models.Db/DbQuestion.cs]: Each option will be connected with only one question, so the navigation property will be Question (not Questions)

File: src/SurveyService.Models.Db/DbQuestion.cs [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+  public bool IsRevoteAvailable { get; set; }
+  public bool IsObligatory { get; set; }
+  public bool IsPrivate { get; set; }
+  public bool HasMultipleChoice { get; set; }
+  public bool HasCustomOptions { get; set; }
+  public Guid CreatedBy { get; set; }
+  public DateTime CreatedAtUTC { get; set; }
+  public Guid? ModifiedBy { get; set; }
+  public DateTime? ModifiedAtUTC { get; set; }
+  public bool IsActive { get; set; }
```
Comment (vorobevamv) [inline src/SurveyService.Models.Db/DbQuestion.cs]: please, move "IsActive" to line 23; Created.. and Modified.. usually are written in the end of Db model

File: src/SurveyService.Models.Db/DbQuestion.cs (hunk 14 above)
Comment (vorobevamv) [inline src/SurveyService.Models.Db/DbQuestion.cs]: "WithOne(o => o.Question)"

File: src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs (hunk 7 above)
Comment (Levliv) [reply src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs]: Multiply - умножать
Multiple - множественный

Comment (maybee16) [reply src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs]: I agree with Lev

File: src/SurveyService.Models.Db/DbQuestion.cs (hunk 14 above)
Comment (maybee16) [reply src/SurveyService.Models.Db/DbQuestion.cs]: Done

File: src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs (hunk 5 above)
Comment (maybee16) [reply src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs]: Done

File: src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs (hunk 6 above)
Comment (maybee16) [reply src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs]: Done

File: src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs (hunk 8 above)
Comment (maybee16) [reply src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs]: Done

File: src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs (hunk 9 above)
Comment (maybee16) [reply src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs]: Done

File: src/SurveyService.Models.Db/DbQuestion.cs (hunk 12 above)
Comment (maybee16) [reply src/SurveyService.Models.Db/DbQuestion.cs]: Done

File: src/SurveyService.Models.Db/DbQuestion.cs (hunk 13 above)
Comment (maybee16) [reply src/SurveyService.Models.Db/DbQuestion.cs]: Done

File: src/SurveyService.Models.Db/DbQuestion.cs (hunk 15 above)
Comment (maybee16) [reply src/SurveyService.Models.Db/DbQuestion.cs]: Done

File: src/SurveyService.Models.Db/DbQuestion.cs (hunk 14 above)
Comment (maybee16) [reply src/SurveyService.Models.Db/DbQuestion.cs]: Done

File: src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs (hunk 7 above)
Comment (vorobevamv) [reply src/SurveyService.Data.Provider.MsSql.Ef/Migrations/20220826180800_InitialQuestionsTable.cs]: OK, by bad, you're right

**Review by vorobevamv — APPROVED** [review]

**Review by Levliv — APPROVED** [review]

File: src/SurveyService.Models.Db/DbQuestion.cs [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+    Options = new HashSet<DbOption>();
+  }
+}
+
+public class QuestionConfiguration : IEntityTypeConfiguration<DbQuestion>
+{
+  public void Configure(EntityTypeBuilder<DbQuestion> builder)
+  {
+    builder
+        .ToTable(DbQuestion.TableName);
```
Comment (Vasis3038) [inline src/SurveyService.Models.Db/DbQuestion.cs]: extra spaces

File: src/SurveyService.Models.Db/DbQuestion.cs [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+
+    builder
+        .HasKey(q => q.Id);
```
Comment (Vasis3038) [inline src/SurveyService.Models.Db/DbQuestion.cs]: extra spaces

File: src/SurveyService.Models.Db/DbQuestion.cs [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+
+    builder
+        .Property(q => q.Content)
+        .IsRequired();
```
Comment (Vasis3038) [inline src/SurveyService.Models.Db/DbQuestion.cs]: extra spaces

File: src/SurveyService.Models.Db/DbQuestion.cs [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+
+    builder
+        .HasMany(q => q.Options)
+        .WithOne(o => o.Question);
+
+    builder
+        .HasOne(q => q.Group)
+        .WithMany(g => g.Questions);
```
Comment (Vasis3038) [inline src/SurveyService.Models.Db/DbQuestion.cs]: extra spaces

File: src/SurveyService.Models.Db/DbQuestion.cs [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+    builder
+        .HasKey(q => q.Id);
+
+    builder
+        .Property(q => q.Content)
+        .IsRequired();
+
+    builder
+        .HasMany(q => q.Options)
+        .WithOne(o => o.Question);
```
Comment (Vasis3038) [inline src/SurveyService.Models.Db/DbQuestion.cs]: extra spaces

File: src/SurveyService.Models.Db/DbQuestion.cs (hunk 16 above)
Comment (maybee16) [reply src/SurveyService.Models.Db/DbQuestion.cs]: Done

File: src/SurveyService.Models.Db/DbQuestion.cs (hunk 18 above)
Comment (maybee16) [reply src/SurveyService.Models.Db/DbQuestion.cs]: Done

File: src/SurveyService.Models.Db/DbQuestion.cs (hunk 19 above)
Comment (maybee16) [reply src/SurveyService.Models.Db/DbQuestion.cs]: Done

File: src/SurveyService.Models.Db/DbQuestion.cs (hunk 20 above)
Comment (maybee16) [reply src/SurveyService.Models.Db/DbQuestion.cs]: Done

File: src/SurveyService.Models.Db/DbQuestion.cs (hunk 17 above)
Comment (maybee16) [reply src/SurveyService.Models.Db/DbQuestion.cs]: Done

File: src/SurveyService.Models.Db/DbQuestion.cs (hunk 20 above)
Comment (AlfaRedDog) [reply src/SurveyService.Models.Db/DbQuestion.cs]: Vasis3038, it's not a mistake, other services have the same syntax

**Review by Vasis3038 — APPROVED** [review]

**Pull request merged**